### PR TITLE
implement both owning and const generic field array writers

### PR DESCRIPTION
**Pull request by burrbull**

### CODE REVIEW

**Comment by burrbull** [comment]
Now we can see something like this:
image

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -25,7 +25,7 @@ jobs:
 
         include:
           # Test MSRV
-          - rust: 1.40.0
+          - rust: 1.51.0
```
Comment (therealprof) [inline .github/workflows/ci.yml:28]: For me requiring 1.51 at this point in time is not ideal. Other opinions, @adamgreig maybe?

Comment (adamgreig) [reply .github/workflows/ci.yml:28]: Is this to generate crates or to use the generated crates? 1.51 is _way_ too new for building PACs, but I wouldn't mind requiring it to generate them.

Comment (therealprof) [reply .github/workflows/ci.yml:28]: Generated crates. It's providing a const generic writer. I would not terribly object if this was hidden behind a feature gate but I fully agree that enabled for all is a non-starter.

Comment (burrbull) [reply .github/workflows/ci.yml:28]: > behind a feature gate

It's idea

Comment (burrbull) [reply .github/workflows/ci.yml]: Done

Comment (burrbull) [reply .github/workflows/ci.yml]: Now this PR doesn't change anything if you don't use `const-generic` feature.

**Comment by burrbull** [comment]
rebased

**Comment by therealprof** [comment]
@adamgreig Any objections to the approach?

**Review by therealprof — CHANGES_REQUESTED** [review]
We should still test this in CI with 1.51.

**Comment by burrbull** [comment]
@therealprof 
I've added test with rust 1.51 and enabled `strict` and `const-generic` features. RISC-V `k210` uses field arrays

**Review by adamgreig — APPROVED** [review]
LGTM

**Review by therealprof — APPROVED** [review]
Thanks a lot!

**Comment by burrbull** [comment]
rebased

**Review by adamgreig — APPROVED** [review]
bors merge

**Pull request merged**